src/goto-instrument/synthesizer/enumerative_loop_invariant_synthesizer.cpp

```diff
@@ -32,19 +86,295 @@ void enumerative_loop_invariant_synthesizert::init_candidates()
       // we only synthesize invariants for unannotated loops
       if(loop_end->condition().find(ID_C_spec_loop_invariant).is_nil())
       {
-        invariant_candiate_map[new_id] = true_exprt();
+        // Store the loop guard.
+        exprt guard =
+          get_loop_head(
+            loop_end->loop_number,
+            goto_model.goto_functions.function_map[function_p.first])
+            ->condition();
+        neg_guards[new_id] = guard;
+
+        // Initialize invariant clauses as `true`.
+        in_invariant_clause_map[new_id] = true_exprt();
+        pos_invariant_clause_map[new_id] = true_exprt();
+      }
+    }
+  }
+}
+
+void enumerative_loop_invariant_synthesizert::build_tmp_post_map()
+{
+  for(auto &goto_function : goto_model.goto_functions.function_map)
+  {
+    for(const auto &instruction : goto_function.second.body.instructions)
+    {
+      // tmp_post variables are symbol lhs of ASSIGN.
+      if(
+        instruction.type() != goto_program_instruction_typet::ASSIGN ||
+        instruction.assign_lhs().id() != ID_symbol)
+        continue;
+
+      const auto symbol_lhs =
+        expr_try_dynamic_cast<symbol_exprt>(instruction.assign_lhs());
+
+      // tmp_post variables have identifiers with the prefix tmp::tmp_post.
+      if(
+        id2string(symbol_lhs->get_identifier()).find("tmp::tmp_post") !=
+        std::string::npos)
+      {
+        tmp_post_map[instruction.assign_lhs()] = instruction.assign_rhs();
+      }
+    }
+  }
+}
+
+// Recursively extract symbols from `expr`
+void extract_symbol_exprt_rec(const exprt &expr, std::set<exprt> &result)
+{
+  if(expr.id() == ID_symbol)
+  {
+    result.insert(expr);
+  }
+  else
+  {
+    for(exprt::operandst::const_iterator it = (expr).operands().begin();
+        it != (expr).operands().end();
+        ++it)
+    {
+      extract_symbol_exprt_rec(*it, result);
+    }
+  }
+}
```

Comment:
Does some function from `util/find_symbols.{h,cpp}` do the trick for you?

Reply:
Yes, thank you!